## Упражнения по библиотеке Numpy

In [40]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [41]:
random_array = np.random.rand(10) * 10
random_array 

array([0.3311411 , 3.00687158, 8.79053135, 3.7435261 , 3.05893704,
       0.85416803, 4.15842382, 9.54149962, 0.18744495, 5.51583218])

In [42]:
# Еще один способ:
#mask = (random_array > 3) & (random_array < 8)
#random_array[mask] = -random_array[mask]
#random_array

random_array = np.where((random_array > 3) & (random_array < 8), -random_array, random_array)
random_array

array([ 0.3311411 , -3.00687158,  8.79053135, -3.7435261 , -3.05893704,
        0.85416803, -4.15842382,  9.54149962,  0.18744495, -5.51583218])

**2.** Заменить максимальный элемент случайного массива на 0

In [43]:
random_array = np.where(random_array == np.max(random_array), 0.0, random_array)
random_array

array([ 0.3311411 , -3.00687158,  8.79053135, -3.7435261 , -3.05893704,
        0.85416803, -4.15842382,  0.        ,  0.18744495, -5.51583218])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [44]:
from itertools import product
arr = np.array([[1, 2], [3, 4], [5, 6]])
result = np.array(list(product(*arr)))
print(result)

[[1 3 5]
 [1 3 6]
 [1 4 5]
 [1 4 6]
 [2 3 5]
 [2 3 6]
 [2 4 5]
 [2 4 6]]


In [45]:
arr = np.array([[1, 2], [3, 4], [5, 6]])

row1 = arr[0]
row2 = arr[1]
row3 = arr[2]

# meshgrid() - создание сетки и 3 матриц строк, stack() - соединение матриц в пары, reshape() - выпрямление результата из 3-мерной матрицу в 2-мерную
res = np.stack(np.meshgrid(row1, row2, row3), axis=-1).reshape(-1, 3)

print(res)

[[1 3 5]
 [1 3 6]
 [2 3 5]
 [2 3 6]
 [1 4 5]
 [1 4 6]
 [2 4 5]
 [2 4 6]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [46]:
A = np.array([
    [1, 2, 3],
    [1, 1, 1],
    [4, 5, 6],
    [2, 4, 9], 
    [0, 0, 0],
    [1, 5, 9],
    [7, 8, 9],
    [3, 3, 3]
])

B = np.array([
    [1, 2],
    [4, 5]
])

mask1 = np.isin(A, B[0]).any(axis=1) 
mask2 = np.isin(A, B[1]).any(axis=1)

final_mask = mask1 & mask2

result = A[final_mask]
print("Подходящие строки из A:\n", result)

Подходящие строки из A:
 [[2 4 9]
 [1 5 9]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [47]:
A = np.random.randint(0, 2, (10, 3))

mask = ~np.all(A == A[:, [0]], axis=1)

result = A[mask]

print("Исходная матрица:\n", A)
print("\nРезультат (только разные значения):\n", result)

Исходная матрица:
 [[1 1 0]
 [1 1 1]
 [1 0 0]
 [0 1 1]
 [1 0 1]
 [1 0 1]
 [0 0 0]
 [1 0 0]
 [1 0 0]
 [1 1 0]]

Результат (только разные значения):
 [[1 1 0]
 [1 0 0]
 [0 1 1]
 [1 0 1]
 [1 0 1]
 [1 0 0]
 [1 0 0]
 [1 1 0]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [48]:
A = np.random.randint(0, 2, (10, 3))

result = np.unique(A, axis=0)

print("Исходная матрица:\n", A)
print("Массив без дубликатов:\n", result)

Исходная матрица:
 [[0 0 0]
 [0 1 1]
 [0 0 0]
 [0 1 0]
 [1 1 1]
 [1 1 1]
 [1 1 1]
 [0 0 1]
 [1 1 1]
 [1 0 1]]
Массив без дубликатов:
 [[0 0 0]
 [0 0 1]
 [0 1 0]
 [0 1 1]
 [1 0 1]
 [1 1 1]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [49]:
matrix = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
min_dimension = min(len(matrix), len(matrix[0]))
result = 1
for i in range(min_dimension):
    if matrix[i][i] != 0:
        result *= matrix[i][i]
print(result)

3


In [50]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
diagonal = X.diagonal()
result = np.prod(diagonal[diagonal != 0])
print(result)

3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [51]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
result = sorted(x) == sorted(y)
print(result)

True


In [52]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
x_sorted = np.sort(x)
y_sorted = np.sort(y)
result = np.array_equal(x_sorted, y_sorted)
print(f"Результат: {result}")

Результат: True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [53]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
max_elem = x[x.index(0) + 1]
for i in range(1, len(x) - 1):
    if x[i - 1] == 0 and x[i] > max_elem:
        max_elem = x[i]
print(max_elem)


5


In [54]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
zero_mask = (x == 0)
elements_after_zeros = x[1:][zero_mask[:-1]]
result = np.max(elements_after_zeros)
print(f"Элементы после нулей: {elements_after_zeros}")
print(f"Ответ: {result}")

Элементы после нулей: [3 0 5]
Ответ: 5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [55]:
x = [2, 2, 2, 3, 3, 3, 5]
nums = []
length = []
i = 0
while i <= len(x)-1:
    nums.append(x[i])
    counter = 1
    j = i + 1
    while j <= len(x)-1 and x[j] == x[i]:
        counter += 1
        j += 1
    length.append(counter)
    i += counter
print(f"nums: {nums} \nlength: {length}")


nums: [2, 3, 5] 
length: [3, 3, 1]


In [56]:
def rle_numpy(x):
    locs = np.where(np.diff(x) != 0)[0]
    pos = np.concatenate([[-1], locs, [len(x) - 1]])
    values = x[pos[1:]]
    counts = np.diff(pos)
    
    return values, counts

x_np = np.array([2, 2, 2, 3, 3, 3, 5])
vals, cnts = rle_numpy(x_np)
print((vals, cnts))

(array([2, 3, 5]), array([3, 3, 1]))


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [57]:
import math
def euclidean_python(X, Y):
    M, K = len(X), len(X[0])
    N = len(Y)
    dist_matrix = [[0.0 for _ in range(N)] for _ in range(M)]
    
    for i in range(M):
        for j in range(N):
            squared_sum = 0
            for k in range(K):
                squared_sum += (X[i][k] - Y[j][k])**2
            dist_matrix[i][j] = math.sqrt(squared_sum)
    return dist_matrix

# Выборка X: 3 объекта, у каждого по 2 координаты (X, Y)
X_py = [
    [0, 0],   # Точка A
    [1, 1],   # Точка B
    [2, 2]    # Точка C
]

# Выборка Y: 2 объекта, с которыми мы хотим сравнить X
Y_py = [
    [1, 0],   # Точка D
    [0, 1]    # Точка E
]

# Результатом будет матрица 3x2 (расстояния A-D, A-E, B-D, B-E...)
result = euclidean_python(X_py, Y_py)
result

[[1.0, 1.0], [1.0, 1.0], [2.23606797749979, 2.23606797749979]]

In [58]:
def euclidean_numpy(X, Y):
    # 1. Сумма квадратов по строкам для X и Y
    # (X**2).sum(axis=1) дает вектор (M,)
    # reshape(-1, 1) делает его столбцом (M, 1) для broadcasting
    x_2 = np.sum(X**2, axis=1).reshape(-1, 1)
    y_2 = np.sum(Y**2, axis=1)
    
    # 2. Матричное произведение X * Y.T (размер M x N)
    xy = X @ Y.T
    
    # 3. Собираем формулу: x^2 - 2xy + y^2
    # Благодаря broadcasting x_2 (столбец) и y_2 (строка) растянутся в матрицу
    dists_sq = x_2 - 2*xy + y_2
    
    # Из-за погрешности float могут появиться крошечные отрицательные числа
    # Ограничим их снизу нулем перед взятием корня
    return np.sqrt(np.maximum(dists_sq, 0))

x_arr = np.array([
    [0, 0],   # Точка A
    [1, 1],   # Точка B
    [2, 2]    # Точка C
])
y_arr = np.array([
    [1, 0],   # Точка D
    [0, 1]    # Точка E
])
print(euclidean_numpy(x_arr, y_arr))


[[1.         1.        ]
 [1.         1.        ]
 [2.23606798 2.23606798]]


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [59]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [60]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [61]:
average_calories = calorie_stats.mean()
average_calories

np.float64(106.88311688311688)

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [62]:
calorie_stats_sorted = sorted(calorie_stats)
calorie_stats_sorted

[np.float64(50.0),
 np.float64(50.0),
 np.float64(50.0),
 np.float64(70.0),
 np.float64(70.0),
 np.float64(80.0),
 np.float64(90.0),
 np.float64(90.0),
 np.float64(90.0),
 np.float64(90.0),
 np.float64(90.0),
 np.float64(90.0),
 np.float64(90.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(100.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(110.0),
 np.float64(1

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [63]:
median_calories = np.median(calorie_stats)
print(f"Median: {median_calories}\nAverage: {average_calories}")

Median: 110.0
Average: 106.88311688311688


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [77]:
percentiles = {}
nth_percentile = [100, np.percentile(calorie_stats, 100)]
for percent in range(1, 99, 1):
    percentiles[percent] = np.percentile(calorie_stats, percent)
    if percentiles[percent] > 60 and percentiles[percent] < nth_percentile[1]:
        nth_percentile = [percent, percentiles[percent]]
#print(percentiles)
print(f"Closest percentile: {nth_percentile[0]}% - {nth_percentile[1]}")


Closest percentile: 4% - 70.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [79]:
more_calories = 100 - nth_percentile[0]
print(f"{more_calories}%")

96%


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [81]:
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

In [80]:
calorie_std = np.std(calorie_stats)
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Калорийность продукта CrunchieMunchies - 60 калорий.
Исходя из представленного разнообразия продуктов на рынке мы можем заметить, что:

1. Средняя калорийность такого продукта на рынке составляет от 86 до 126 калорий. Больше половины из них превышают 110 калорий
2. Продукт CrunchieMunchies менее калорийный, чем 96% полезных злаков, представленных на рынке
3. Продукт CrunchieMunchies можно смело считать диетическим и делать маркетинговый упор на то, что он подходит как для всех, так и для тех, кто следит за своей фигурой